AmplifyPlugins/API/Sources/AWSAPIPlugin/Operation/AWSGraphQLSubscriptionTaskRunner.swift

```diff
@@ -402,5 +402,36 @@ fileprivate func toAPIError<R: Decodable>(_ errors: [Error], type: R.Type) -> AP
             errors.first
         )
     }
+}
+
+fileprivate func decodeAppSyncRealTimeResponseError(_ data: JSONValue?) -> [Error] {
+    let knownAppSyncRealTimeRequestErorrs =
+        decodeAppSyncRealTimeRequestError(data)
+        .filter { !$0.isUnknown }
+    if knownAppSyncRealTimeRequestErorrs.isEmpty {
+        let graphQLErrors = decodeGraphQLErrors(data)
+        return graphQLErrors.isEmpty
+            ? [APIError.operationError("Failed to decode AppSync error response", "", nil)]
+            : graphQLErrors
+    } else {
+        return knownAppSyncRealTimeRequestErorrs
+    }
+}
+
+fileprivate func decodeGraphQLErrors(_ data: JSONValue?) -> [GraphQLError] {
+    do {
+        return try GraphQLErrorDecoder.decodeAppSyncErrors(data)
+    } catch {
+        print("Failed to decode errors: \(error)")
+        return []
+    }
+}
 
+fileprivate func decodeAppSyncRealTimeRequestError(_ data: JSONValue?) -> [AppSyncRealTimeRequest.Error] {
+    guard let errorsJson = data?.errors else {
+        print("No 'errors' field found in response json")
+        return []
+    }
+    let errors = errorsJson.asArray ?? [errorsJson]
+    return errors.compactMap(AppSyncRealTimeRequest.parseResponseError(error:))
 }
```

Comment:
should we move these under an extension of something and test them? functions at the top level of the file seems a bit out of place.

Reply:
sure, will make it internal static method and add unit test cases.